### PR TITLE
Add siddhi language server implementation

### PR DESCRIPTION
**Pull request by mohanvive**
by @Malintha1996

### CODE REVIEW

**Comment by mohanvive** [comment]
@Malintha1996 Do you have any update on this ?

**Comment by dnwick** [comment]
@Malintha1996 are we done with all the changes ? what are the pending tasks ? And have you added any improvements after the code review ?

**Comment by Malintha1996** [comment]
Hello @dnwick, No, there are few other pending tasks to be completed. I will be committing changes as small chunks for the convenience of reviewing. The most recent commit concludes the work related to snippet generation.I will upload a full report of changes that have to be done soon.

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/definitionwindow/DefinitionWindowContextProvider.java [hunk 1]
```diff
@@ -40,13 +41,13 @@ public DefinitionWindowContextProvider() {
...
 
-        List<Object[]> suggestions = new ArrayList<>();
+        List<Map<String, Object>> suggestions = new ArrayList<>();
         ParserRuleContext definitionWindowContext = (ParserRuleContext) LSOperationContext.INSTANCE.getParseTreeMap()
                 .get(SiddhiQLParser.Definition_windowContext.class.getName());
         int childCount = definitionWindowContext.getChildCount();
         if (childCount > 0) {
             if (definitionWindowContext.getChild(childCount - 1) instanceof SiddhiQLParser.SourceContext) {
-                suggestions.add(SnippetBlockUtil.ATTRIBUTE_LIST_SNIPPET);
+                suggestions.add(SnippetBlockUtil.SNIPPETS.get("TRIBUTE_LIST_SNIPPET"));
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/definitionwindow/DefinitionWindowContextProvider.java]: is this a typo .. TRIBUTE ?

File: runner/components/io.siddhi.langserver/pom.xml [hunk 2]
```diff
@@ -58,10 +63,11 @@
     <properties>
         <lsp4j.version>0.8.1</lsp4j.version>
         <antlr4-runtime.version>4.7.2</antlr4-runtime.version>
-        <siddhi-core.version>5.1.8</siddhi-core.version>
-        <siddhi-query-compiler.version>5.1.8</siddhi-query-compiler.version>
+        <siddhi-core.version>5.1.12-SNAPSHOT</siddhi-core.version>
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/pom.xml]: dont send snapshot versions in ur PR

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/CompletionProvider.java [hunk 3]
```diff
@@ -62,20 +62,26 @@ public ParserRuleContext getParent() {
...
-                    CompletionItem completionItem = new CompletionItem();
-                    completionItem.setInsertText((String) suggestion[0]);
-                    completionItem.setLabel((String) suggestion[1]);
-                    completionItem.setKind((CompletionItemKind) suggestion[2]);
-                    completionItem.setDetail((String) suggestion[3]);
-                    completionItem.setFilterText((String) suggestion[4]);
-                    completionItem.setInsertTextFormat((InsertTextFormat) suggestion[5]);
-                    completionItems.add(completionItem);
+            for (Map<String, Object> suggestion : suggestions) {
+                if (suggestion.size() == 6) {
```
[the remark this answers is not in the document]
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/CompletionProvider.java]: done

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/CompletionProvider.java [hunk 4]
```diff
@@ -62,20 +62,26 @@ public ParserRuleContext getParent() {
...
+                    try {
+                        CompletionItem completionItem = new CompletionItem();
+                        completionItem.setInsertText((String) suggestion.get("insertText"));
```
[the remark this answers is not in the document]
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/CompletionProvider.java]: done

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/definitionwindow/DefinitionWindowContextProvider.java (hunk 1 above)
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/definitionwindow/DefinitionWindowContextProvider.java]: done

File: runner/components/io.siddhi.langserver/pom.xml (hunk 2 above)
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/pom.xml]: done

**Comment by Malintha1996** [comment]
@dnwick I have made the requested changes.

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/SiddhiTextDocumentService.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * TextDocumentService implementation for Siddhi provide various services
+ * {@link TextDocumentService} upon clients' requests.
+ */
+public class SiddhiTextDocumentService implements TextDocumentService {
+
+    // todo:indicates the frequency to generate diagnostics upon document did change
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/SiddhiTextDocumentService.java:47]: can u remove these todo s and comments please

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/SiddhiTextDocumentService.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        this.diagnosticProvider = LSOperationContext.INSTANCE.getDiagnosticProvider();
+    }
+
+    /**
+     * Set the client's capabilities, this is used for concurrent operation of language server which is not
+     * implemented currently.
+     *
+     * @param clientCapabilities Client's Text Document Capabilities.
+     */
+//    public void setClientCapabilities(TextDocumentClientCapabilities clientCapabilities) {
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/SiddhiTextDocumentService.java:67]: remove these.

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/OrderByReferenceContextProvider.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+
+/**
+ * Provides completions for OrderByReferenceContext.
+ * {@link io.siddhi.query.compiler.SiddhiQLParser.Order_by_referenceContext}.
+ */
+public class OrderByReferenceContextProvider extends CompletionProvider {
+
+    public OrderByReferenceContextProvider() {
+
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/OrderByReferenceContextProvider.java:34]: remove new lines

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/OrderByReferenceContextProvider.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        this.providerName = SiddhiQLParser.Order_by_referenceContext.class.getName();
+    }
+
+    @Override
+    public List<CompletionItem> getCompletions() {
+
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/OrderByReferenceContextProvider.java:40]: same here

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/SourceContextProvider.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+public class SourceContextProvider extends CompletionProvider {
+
+    public SourceContextProvider() {
+
+        this.providerName = SiddhiQLParser.SourceContext.class.getName();
+    }
+
+    @Override
+    public List<CompletionItem> getCompletions() {
+
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/completion/providers/common/SourceContextProvider.java:44]: remove

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/AttributeMetaData.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/AttributeMetaData.java:2]: wrong license header

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/Constants.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/Constants.java:2]: same here

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/visitor/LanguageServerParserErrorStrategy.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        this.recognizer = recognizer;
+        generateContextTree();
+        String msg = "Mismatched input " + this.getTokenErrorDisplay(e.getOffendingToken()) + " expecting " +
+                e.getExpectedTokens().toString(recognizer.getVocabulary());
+        recognizer.notifyErrorListeners(e.getOffendingToken(), msg, e);
+    }
+
+    @Override
+    protected void reportFailedPredicate(Parser recognizer, FailedPredicateException e) {
+
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/visitor/LanguageServerParserErrorStrategy.java:120]: remove new lines

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/visitor/LanguageServerParserErrorStrategy.java [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            Token token = recognizer.getCurrentToken();
+            IntervalSet expecting = this.getExpectedTokens(recognizer);
+            String msg =
+                    "Missing " + expecting.toString(recognizer.getVocabulary()) + " at " +
+                            this.getTokenErrorDisplay(token);
+            recognizer.notifyErrorListeners(token, msg, null);
+        }
+    }
+
+    //todo: add this method in the case of goalPosition is placed after the erroneous position in the source code.
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/visitor/LanguageServerParserErrorStrategy.java:160]: remove or fix it

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/test/java/io/siddhi/langserver/utils/FileUtil.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
+package io.siddhi.langserver.utils;
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/test/java/io/siddhi/langserver/utils/FileUtil.java:1]: no license

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/test/java/io/siddhi/langserver/utils/TestUtil.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
+package io.siddhi.langserver.utils;
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/test/java/io/siddhi/langserver/utils/TestUtil.java:1]: no license

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/main/java/StdioLauncher.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+public class StdioLauncher {
+
+    /**
+     * Main method is executed at the classpath to start Siddhi language server.
+     *
+     * @param args language server/client options
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    public static void main(String args[]) throws InterruptedException, ExecutionException {
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/main/java/StdioLauncher.java:39]: do u want this main class ?

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/test/resources/testng.xml [hunk 17]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/test/resources/testng.xml:3]: wrong license

File: runner/components/io.siddhi.langserver/pom.xml [hunk 18]
```diff
@@ -0,0 +1,66 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (dnwick) [inline runner/components/io.siddhi.langserver/pom.xml:1]: add license

**Comment by dnwick** [comment]
@Malintha1996 there are several license issues , some are wrong and some are missing and there are unnecessary new lines and commented out code. Please go through each file and fix those

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/main/java/StdioLauncher.java (hunk 16 above)
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.launcher/src/main/java/StdioLauncher.java:39]: > do u want this main class ?

Yes. This is the general structure that has been followed by other language server implementations as well. A sperate launcher class is used for launching the language server-client connection.  For instance,  ballerina launcher implementation is as follows
https://github.com/ballerina-platform/ballerina-lang/blob/a1fd9f45de540387ca6ad687f7d8c2b625b830e1/language-server/modules/launchers/stdio-launcher/src/main/java/org/ballerinalang/langserver/launchers/stdio/Main.java#L32-L48

**Comment by Malintha1996** [comment]
> @Malintha1996 there are several license issues , some are wrong and some are missing and there are unnecessary new lines and commented out code. Please go through each file and fix those

Done!

File: runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/ParameterMetaData.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/ParameterMetaData.java:2]: > @Malintha1996 Did u went through my all comments i still see some files have same issues ??

My bad. I have missed them. I will fix those.

[the remark this answers is not in the document]
Comment (Malintha1996) [reply runner/components/io.siddhi.langserver/io.siddhi.langserver.core/src/main/java/io/siddhi/langserver/utils/metadata/ParameterMetaData.java]: > same here ???

done. This PR doesn't have the implementation to add language server jars to plugins/libs folder. Therefore please don't merge the PR yet. I will make the necessary implementations and commit asap

**Pull request closed without merging**